test/Constraints/keypath.swift

```diff
@@ -19,3 +19,15 @@ func test() {
 
   let _ = C()[keyPath: \.i] // no warning for a read
 }
+
+// SR-7339
+class Some<T, V> {
+  init(keyPath: KeyPath<T, ((V) -> Void)?>) {
+  }
+}
+
+class Demo {
+  var here: (() -> Void)?
+}
+let some = Some(keyPath: \Demo.here)
+// expected-error@-1 {{cannot convert value of type 'ReferenceWritableKeyPath<Demo, (() -> Void)?>' to expected argument type 'KeyPath<_, ((_) -> Void)?>'}}
```

Comment:
Could you please add a `FIXME` here to improve this diagnostic, it's better in a way that points to a correct place but doesn't really pinpoint the problem unfortunately still...